flow/workflows/cancel_table_addition_flow.go

```diff
@@ -0,0 +1,280 @@
+package peerflow
+
+import (
+	"fmt"
+	"time"
+
+	"go.temporal.io/sdk/temporal"
+	"go.temporal.io/sdk/workflow"
+
+	"github.com/PeerDB-io/peerdb/flow/generated/protos"
+	"github.com/PeerDB-io/peerdb/flow/workflows/cdc_state"
+)
+
+func CancelTableAdditionFlow(ctx workflow.Context, input *protos.CancelTableAdditionInput) (*protos.CancelTableAdditionOutput, error) {
+	logger := workflow.GetLogger(ctx)
+	flowJobName := input.FlowJobName
+
+	logger.Info("Starting cancel table addition flow", "flowName", flowJobName)
+	getFlowConfigOptions := workflow.ActivityOptions{
+		StartToCloseTimeout: time.Minute * 5,
+		RetryPolicy: &temporal.RetryPolicy{
+			InitialInterval:    time.Second * 5,
+			BackoffCoefficient: 2.0,
+			MaximumInterval:    time.Minute * 1,
+			MaximumAttempts:    5,
+		},
+	}
+	getFlowConfigCtx := workflow.WithActivityOptions(ctx, getFlowConfigOptions)
+
+	var flowConfigFetchOutput *protos.GetFlowConfigAndWorkflowIdFromCatalogOutput
+	err := workflow.ExecuteActivity(getFlowConfigCtx,
+		cancelTableAddition.GetFlowConfigAndWorkflowIdFromCatalog, flowJobName).Get(ctx, &flowConfigFetchOutput)
+	if err != nil {
+		logger.Error("Failed to get flow config from catalog", "error", err)
+		return nil, err
+	}
+	flowConfig := flowConfigFetchOutput.FlowConnectionConfigs
+	originalWorkflowId := flowConfigFetchOutput.WorkflowId
+	logger.Info("Retrieved flow config", "flowName", flowJobName, "tableCount", len(flowConfig.TableMappings))
+
+	currentlyReplicatingTableSet := make(map[string]bool)
+	for _, mapping := range input.CurrentlyReplicatingTables {
+		currentlyReplicatingTableSet[mapping.SourceTableIdentifier] = true
+	}
+
+	var removedTables []*protos.TableMapping
+	for _, mapping := range flowConfig.TableMappings {
+		if !currentlyReplicatingTableSet[mapping.SourceTableIdentifier] {
+			removedTables = append(removedTables, mapping)
+		}
+	}
+
+	if len(removedTables) > 0 {
+		if !input.AssumeTableRemovalWillNotHappen {
+			return nil, fmt.Errorf("cannot cancel table addition because the following tables are expected to be removed during the operation and cleanup for that is not implemented: %v; "+
```

Comment:
The caller workflow would need to handle this error and error out without retrying forever. Maybe introduce a well-known error, match it on the API side and return a PreconditionFailed?